README.md

```diff
@@ -1,14 +1,124 @@
+remotedialer creates a two-way connection between two entities which allows
+services running on localhost or behind a NAT or firewall to be accessible by a
+remote server.
 
-Fun times!
+Architecture
+------------
 
-Refer to [`server/`](server/) and [`client/`](client/) how to use.  Or don't.... This framework can hurt your head
-trying to conceptualize.
+### Abstractions
 
-See also:
+remotedialer consists of structs that organize and abstract a TCP connection
+between a server and a client. Both client and server use ``Session``s, which
+provides a means to make a network connection to a remote endpoint and keep
+track of incoming connections. A server uses the ``Server`` object which
+contains a ``sessionManager`` which governs one or more ``Session``s, while a
+client creates a ``Session`` directly. A ``connection`` implements the
+``io.Reader`` and ``io.WriteCloser`` interfaces, so it can be read from and
+written to directly. The ``connection``'s internal ``readBuffer`` monitors the
+size of the data it is carrying and uses ``backPressure`` to pause incoming
+data transfer until the amount of data is below a threshold.
 
-* [inlets.dev](https://inlets.dev) which uses the client and server components to form a tunnel for clients behind NAT or firewalls.
+![](./docs/remotedialer.png)
+
+### Data flow
+
+![](./docs/remotedialer-flow.png)
+
+A client establishes a session with a server using the server's URL. The server
+upgrades the connection to a websocket connection which it uses to create a
+``Session``. The client then also creates a ``Session`` with the websocket
+connection.
+
+The client sits in front of some kind of HTTP server, most often a kubernetes
+API server, and acts as a reverse proxy for that HTTP resource. When a user
+requests a resource from this remote resource, request first goes to the
+remotedialer server. The application containing the server is responsible for
+routing the request to the right client connection.
+
+The request is sent through the websocket connection to the client and read
+into the client's connection buffer. A pipe is created between the client and
+the HTTP service which continually copies data between each socket. The request
+is forwarded through the pipe to the remote service, draining the buffer. The
+service's response is copied back to the client's buffer, and then forwarded
+back to the server and copied into the server connection's own buffer. As the
+user reads the response, the server's buffer is drained.
+
+The pause/resume mechanism checks the size of the buffer for both the client
+and server. If it is greater than the threshold, a ``PAUSE`` message is sent
+back to the remote connection, as a suggestion not to send any more data. As
+the buffer is drained, either by the pipe to the remote HTTP service or the
+user's connection, the size is checked again. When it is lower than the
+threshold, a ``RESUME`` message is sent, and the data transfer may continue.
```

Comment:
Is there anything preventing both the server and client's from entering pause at the same time? Or are all the requests processed serially to where that won't happen naturally?

Reply:
There's no serialization afaict, so if both server and client happen to send pause messages at the same time they would each go into pause mode.

Reply:
This doesn't matter right since the buffers can still be drained in this case it won't lead to a deadlock?

Reply:
Right, since reading and writing are separate, when a connection is paused it can be no longer written to, but the other side can continue to read from it and drain it and eventually cause it to resume.